### PR TITLE
Port Mapping Rule List

### PR DESCRIPTION
**Pull request by Lan-Hekary**
I tested the new branch Expensively on three different routers here with me plus one router from @lucasromeiro 
I made it faster .. so fast that you can printAllPortMappings(); of the router in under a second :)
and Add Multiple rules and Verify them in order ..
this optimization was due to new method in searching the response and quickly flushing the remaining data to allow for the new request to be parsed ..

I will Pull Request the Branch now .. but I ask you not to merge it until we do a heavy testing and try to Enrich the API more ..
I added a Method to get the External IP Address and I extracted the remaining Lease Duration too ..

### CODE REVIEW

**Comment by ofekp** [comment]
Thank you, will review the change in few days. Meanwhile, please look at the link I gave you in the previous PR which explains why the use of String is not recommended. We should think of better ways to optimize the memory usage.
I will like to know which TLS package you're using so that I can experiment with it and optimize the code.
Thank you.

**Comment by Lan-Hekary** [comment]
the TLS Package I am using is BeerSSL ..
I am using the staging version of :
https://github.com/esp8266/Arduino

I read the article at :
https://miscsolutions.wordpress.com/2011/10/16/five-things-i-never-use-in-arduino-projects/

and yes he is right about "Dynamic memory allocation typically causes memory fragmentation"
but also he did mention the Arduino Platform .. not the ESP8266 or ESP32 ..

ESP8266 Handles Dynamic memory allocation differently and Most if the internal workings of the ESP use Dynamic memory allocation extensively ..
I also measured the fragmentation Before and after the Library Methods are called .. typically the fragmentation is at the first call due to the allocation of the WiFiClient and the UDPClient ..
but after that .. every call does not change it a little bit ..
the whole system worked for 8 hours .. updating the UPnP every 10 mins and serving HTTP Server plus some websockets ..
I'd say it's Pretty stable :)
but I will keep stress testing it in the mean while ..
and I'd like you to stress test it too .. and find it's limits ..

thanks for the article by the way .. Good Stuff :D

**Comment by lucasromeiro** [comment]
Hello,
I can help more in the tests.
I have a product running on some clients and I can use this library.
But I can only do this when we have a version with all possible improvements and improvements. Because I can not cause many problems for customers.
On one of my routers we could not get it to work.
So there are still some things to improve.
Also what you have already talked about here and other topics.
The library has improved a lot !!
I must congratulate everyone on the effort.
I'd like to help more in coding, but I'm full of work.
But you can count on me for the tests !!
When we put together all these details and launch a version with all the fixes I think I can by in some clients and test the operation 24 hours a day!

File: src/TinyUPnP.cpp [hunk 1]
```diff
@@ -484,51 +438,345 @@ boolean TinyUPnP::waitForUnicastResponseToMSearch(gatewayInfo *deviceInfo, IPAdd
...
+
+IPAddress TinyUPnP::getExternalIP(){
+	IPAddress myIP; 
+	_upnpRule *result_ptr = new _upnpRule();
+	upnpResult result = postSOAPAction(&_gwInfo, result_ptr, GetExternalIPAddress);
+	if (result != UPNP_OK){
+		UPNP_DEBUG("Error:");
+		UPNP_DEBUGln(String(result));
+		delete result_ptr;
+		return false;
```
Comment (ofekp) [inline src/TinyUPnP.cpp]: :/ return type

Comment (Lan-Hekary) [reply src/TinyUPnP.cpp]: oh, Totally missed it .. thanks, I'll change it ASAP ..
also don't mind the Commented parts of the code .. I will remove them for the final Pull but after complete testing ..

File: src/TinyUPnP.cpp [hunk 2]
```diff
@@ -960,45 +1078,33 @@ String TinyUPnP::getPath(String url) {
...
+	UPNP_DEBUG(line);
+	UPNP_DEBUG(F("] for end tag [</"));
+	UPNP_DEBUG(tagName);
+	UPNP_DEBUGln(F(">]"));
     return "";
   }
   return line.substring(startIndex, endIndex);
-}
-
-// prints a message to Serial only if IS_DEBUG is set to true
```
Comment (ofekp) [inline src/TinyUPnP.cpp]: ✔️

File: src/TinyUPnP.cpp [hunk 3]
```diff
@@ -960,45 +1078,33 @@ String TinyUPnP::getPath(String url) {
...
+String TinyUPnP::getTagContent(String &line, String tagName, int startIndex) {
   if (startIndex == -1) {
-	debugPrint(F("ERROR: Cannot find tag content in line ["));
-	debugPrint(line);
-	debugPrint(F("] for start tag [<"));
-	debugPrint(tagName);
-	debugPrintln(F(">]"));
+		startIndex = line.indexOf("<" + tagName + ">");
+	}
+  if (startIndex == -1) {
```
Comment (ofekp) [inline src/TinyUPnP.cpp]: not sure I get what's going on here.

Comment (Lan-Hekary) [reply src/TinyUPnP.cpp]: I didn't change this bit of code .. it's the same as the old one ..
the lines numbers changed ..

**Comment by Lan-Hekary** [comment]
> Please work with me here, I am going to review this, you've done some great things, but I will not allow the Strings that are generated on every attempt to set the rules. Other heap usage such as the rules nodes I will allow since it is generated once.
> Please understand my mindset here.
> I will review everything and will get inside as much as I can. Will add comments and will create some changes of my own.
> Will take me some time. Please be patient.

if you mean the :
`String request` at the line 493 
then I'd like you to propose a code that will allow the generation of the request dynamically without using Strings ..
the only idea in my mind is to send the char to the client and remove the Content-length header and replace it with a Transfer-Encoding: chunked
but I don't use this method normally ..

**Comment by ofekp** [comment]
Please explain this line to me:
`boolean urlBaseFound = true;		// don't look for it`
and:
`if (deviceInfo->host == host){ // check for multiple routers`

Still going over this, but need you to fill me up on this please.
Thanks.

**Comment by ofekp** [comment]
@Lan-Hekary 
https://github.com/Lan-Hekary/TinyUPnP/pull/1

**Pull request merged**